File: src/main/scala/com/gitlab/dhorman/cryptotrader/util/Secrets.scala
package com.gitlab.dhorman.cryptotrader.util

import java.nio.charset.StandardCharsets
import java.nio.file.{Files, Paths}

trait Secrets {
  def get(secretId: String): Option[String]
}

object Secrets extends Secrets {
  private val secrets = new NoSecrets with DockerSecrets with EnvBasedSecrets

  override def get(secretId: String): Option[String] = secrets.get(secretId)


  private trait EnvBasedSecrets extends Secrets {
    abstract override def get(secretId: String): Option[String] = {
      Option(System.getenv(secretId)).orElse({super.get(secretId)})
    }
  }

  private trait FileBasedSecrets extends Secrets {
    protected def getPathPrefix: String

    abstract override def get(secretId: String): Option[String] = try {
      val filePath = Paths.get(getPathPrefix + secretId)
      if (!Files.exists(filePath)) return super.get(secretId)
      val contents = Files.readAllBytes(filePath)
      Some(new String(contents, StandardCharsets.UTF_8))
    } catch {
      case _: Exception => super.get(secretId)
    }
  }

  private trait DockerSecrets extends FileBasedSecrets {
    override protected def getPathPrefix = "/run/secrets/"
  }

  private class NoSecrets extends Secrets {
    override def get(secretId: String): Option[String] = None
  }
}

File: trading-robot/src/main/kotlin/com/gitlab/dhorman/cryptotrader/util/Secrets.kt
package com.gitlab.dhorman.cryptotrader.util

import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Paths

interface ISecrets {
    fun get(secretId: String): String?
}

object Secrets : ISecrets {
    private val secrets = EnvBasedSecrets(DockerSecrets(NoSecrets()))

    override fun get(secretId: String): String? = secrets.get(secretId)
}

private class EnvBasedSecrets(private val secret: ISecrets) : ISecrets {
    override fun get(secretId: String): String? {
        return System.getenv(secretId) ?: secret.get(secretId)
    }
}

private abstract class FileBasedSecrets(private val secret: ISecrets) : ISecrets {
    protected abstract fun getPathPrefix(): String

    override fun get(secretId: String): String? {
        try {
            val filePath = Paths.get(getPathPrefix() + secretId)
            if (!Files.exists(filePath)) return secret.get(secretId)
            val contents = Files.readAllBytes(filePath)
            return String(contents, StandardCharsets.UTF_8)
        } catch (_: Exception) {
            return secret.get(secretId)
        }
    }
}

private class DockerSecrets(secret: ISecrets) : FileBasedSecrets(secret) {
    override fun getPathPrefix() = "/run/secrets/"
}

private class NoSecrets : ISecrets {
    override fun get(secretId: String): String? = null
}
